### PR TITLE
[JENKINS-36184]  Doing reflection with hudson.matrix.MatrixRun

### PR DESCRIPTION
**Pull request by fbelzunc**
This is to avoid issues like [JENKINS-36184](https://issues.jenkins-ci.org/browse/JENKINS-36184) 

```
jun 23, 2016 1:24:01 PM hudson.model.Executor finish1
SEVERE: Executor threw an exception
java.lang.NoClassDefFoundError: hudson/matrix/MatrixRun
    at org.jenkinsci.lib.envinject.service.EnvInjectActionRetriever.getEnvInjectAction(EnvInjectActionRetriever.java:25)
    at org.jenkinsci.lib.envinject.service.EnvVarsResolver.getEnVars(EnvVarsResolver.java:56)
    at hudson.plugins.batch_task.BatchRun.run(BatchRun.java:224)
    at hudson.model.ResourceController.execute(ResourceController.java:98)
    at hudson.model.Executor.run(Executor.java:410)
Caused by: java.lang.ClassNotFoundException: hudson.matrix.MatrixRun
    at jenkins.util.AntClassLoader.findClassInComponents(AntClassLoader.java:1376)
    at jenkins.util.AntClassLoader.findClass(AntClassLoader.java:1326)
    at jenkins.util.AntClassLoader.loadClass(AntClassLoader.java:1079)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
    ... 5 more
```

@reviewbybees CC @oleg-nenashev

### CODE REVIEW

**Comment by alvarolobato** [comment]
🐝 provided it works.

**Comment by amuniz** [comment]
This is all really weird.

This is not a Jenkins plugin (at least it is not using the `hpi` packaging).

So any plugin using this lib as dependency directly is going to miss its tansitive plugin dependencies (they are **not** going to be installed in Jenkins). I don't think `matrix-project` is going to be installed (as it seems to be the intention of this PR).

However, I've found that there is an [envinject-plugin](https://github.com/jenkinsci/envinject-plugin) that uses this lib (may a wrapper for it with UI? @oleg-nenashev seems to be a maintainer). If my understanding is correct, the right fix JENKINS-36184 is to add `envinject`-plugin (not lib) as dependency of `batch-task-plugin`.

**Comment by alvarolobato** [comment]
@amuniz if the plugin is a wrapper, you are correct. That concern is what I was saying here: https://github.com/jenkinsci/batch-task-plugin/pull/6

MatrixProject is bundled and cannot be uninstalled so we should be fine. In any case I would move the dependency to the plugin and add envinject-plugin to batch-task-plugin removing the lib.

**Comment by oleg-nenashev** [comment]
It looks like a hack in any case (even if you put it into EnvInject).
Jenkins Core is supposed to resolved such transient dependencies.

This fix is not going to fly 🐛

**Comment by fbelzunc** [comment]
@oleg-nenashev @alvarolobato Would you mind to review, please? I think this is the right fix.

**Comment by oleg-nenashev** [comment]
It's better than the previous fix, but it still depends on how JIT links the code. All MatrixRun usages should be moved to a separate method.

**Comment by fbelzunc** [comment]
@oleg-nenashev My opinion is that this fix should go in. We cannot wait until JIT gets fixed.

**Cannot we merge and release this?** batch-task plugin and later we will have other plugin will be totally broken because of this.

**Comment by oleg-nenashev** [comment]
@fbelzunc JIT behaves correctly. You just should have code in separate method and check LinkageError outside the method. IIRC

**Comment by alvarolobato** [comment]
@fbelzunc @oleg-nenashev  Isn't it easier to just do that single line of code with reflection and avoid having a hard reference to MatrixRun and all the JIT nightmare?

@fbelzunc  I think that if you do it the way I suggested this morning you will be fine.

**Comment by alvarolobato** [comment]
I would suggest something in this line:
```
       List<Action> actions;
        try {
            Class<?> matrix = Class.forName("hudson.matrix.MatrixRun");
            if (matrix.isInstance(build)) {
                build=(AbstractBuild<?, ?>)(matrix.getMethod("getParentBuild").invoke(build));
            }
        } catch (ClassNotFoundException e) {
            //just ignore it, Matrix not available
        } catch (Exception e) {
            //Just log the error, and it will fall back to previous behaviour...
        }
        
        actions = build.getActions();
       ...
```

**Comment by fbelzunc** [comment]
@alvarolobato Many thanks, I was doing a stupid thing.

@oleg-nenashev I know you felt I was pushing you a lot when it is not a responsibility of a plugin maintainer that the PR submitter disappeared for a week. I think you are right and I hope you take my apologizes on this behaviour.

**Comment by amuniz** [comment]
@fbelzunc if you are going to copy&paste @alvarolobato's code, please don't copy the code style horror also 😄 

BTW the approach sounds good to me.

**Comment by fbelzunc** [comment]
@amuniz what do you mean by code style horror?

**Comment by alvarolobato** [comment]
@amuniz 😠

File: src/main/java/org/jenkinsci/lib/envinject/service/EnvInjectActionRetriever.java [hunk 1]
```diff
@@ -16,15 +19,28 @@
...
+            Class<?> matrixClass = Class.forName("hudson.matrix.MatrixRun");
+            if (matrixClass.isInstance(build)) {
+                Method method = matrixClass.getMethod("getParentBuild", null);
+                Object object = method.invoke(build);
+                if (object instanceof AbstractBuild<?, ?>) {
+                    build = (AbstractBuild<?, ?>) object;
+                }
+            }
+        } catch (ClassNotFoundException e) {
+            LOGGER.log(Level.FINEST, String.format("hudson.matrix.MatrixRun is missed", e));
```
Comment (alvarolobato) [inline src/main/java/org/jenkinsci/lib/envinject/service/EnvInjectActionRetriever.java]: I would say: "hudson.matrix.MatrixRun is not installed"

File: src/main/java/org/jenkinsci/lib/envinject/service/EnvInjectActionRetriever.java [hunk 2]
```diff
@@ -16,15 +19,28 @@
     //with EnvInjectAction subclasses. Subclasses cannot be casted from
     //all point of Jenkins (classes are not loaded in some points)
     public Action getEnvInjectAction(AbstractBuild<?, ?> build) {
-
+        List<Action> actions;
```
Comment (alvarolobato) [inline src/main/java/org/jenkinsci/lib/envinject/service/EnvInjectActionRetriever.java]: You could keep this bellow where it was. If build is null you don't need this variable.

File: src/main/java/org/jenkinsci/lib/envinject/service/EnvInjectActionRetriever.java [hunk 3]
```diff
@@ -16,15 +19,28 @@
...
+        } catch (ClassNotFoundException e) {
+            LOGGER.log(Level.FINEST, String.format("hudson.matrix.MatrixRun is missed", e));
+        } catch (NoSuchMethodException e) {
+            LOGGER.log(Level.WARNING, String.format("The method getParentBuild does not exist for hudson.matrix.MatrixRun", e));
+        } catch (IllegalAccessException e) {
+            LOGGER.log(Level.WARNING, String.format("There was a problem in the invocation of getParentBuild in hudson.matrix.MatrixRun", e));
+        } catch (InvocationTargetException e) {
+            LOGGER.log(Level.WARNING, String.format("There was a problem in the invocation of getParentBuild in hudson.matrix.MatrixRun", e));
+        } finally {
             actions = build.getActions();
```
Comment (alvarolobato) [inline src/main/java/org/jenkinsci/lib/envinject/service/EnvInjectActionRetriever.java]: 🐛 I would take this out of the finally. If you have a RuntimeException you won't want to execute this statement.

**Comment by alvarolobato** [comment]
🐝

**Comment by oleg-nenashev** [comment]
🐝 
@reviewbybees done
Merging as a maintainer

**Pull request merged**